Add optional attribute to spec

criccomini opened the pull request:
The current Recap null-field handling is a bit cumbersome. I discussed with
@gwukelic, @adrianisk, and @cpard. We decided to add support for an `optional`
field to the CST. This will allow us to specify that a field is optional.

See the conversation here:

https://github.com/recap-build/recap/pull/335

Users can now specify an optional field as:

```yaml
type: struct
fields:
  - name: secondary_phone
    type: string32
    optional: true
```

This will get treated the same as:

```yaml
type: struct
fields:
  - name: secondary_phone
    type: "union",
    types: ["null", "string32"],
    default: null
```

criccomini committed:
Add TOC

criccomini committed:
Add `optional` attribute to spec

The current Recap null-field handling is a bit cumbersome. I discussed with
@gwukelic, @adrianisk, and @cpard. We decided to add support for an `optional`
field to the CST. This will allow us to specify that a field is optional.

See the conversation here:

https://github.com/recap-build/recap/pull/335

Users can now specify an optional field as:

```yaml
type: struct
fields:
  - name: secondary_phone
    type: string32
    optional: true
```

This will get treated the same as:

```yaml
type: struct
fields:
  - name: secondary_phone
    type: "union",
    types: ["null", "string32"],
    default: null
```

gwukelic commented on specs/type/0.2.0.md:
```diff
@@ -0,0 +1,747 @@
+# A union type of null or a 32-bit signed int
+type: union
+types:
+  - type: null
```
Does this work now? I'm worried that `null` (without quotes) still won't work. If so, we should update the examples to use `"null"` (with quotes).

criccomini replied:
Ah, good catch. Updated.

criccomini committed:
Replace `null` with `"null"` in YAML examples

YAML coerces `null` to a null literal (None in Python) rather than a string
literal ("null"). The examples should say `"null"` since all type attribues in
Recap have a string literal value.
